Custom op autograd tests

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30927 Fix missing virtual destructor
* **#30519 Custom op autograd tests**

Re-enable them and write a few additional ones

Differential Revision: [D18729561](https://our.internmc.facebook.com/intern/diff/D18729561/)

smessmer committed:
Custom op autograd tests

Re-enable them and write a few additional ones

Differential Revision: [D18729561](https://our.internmc.facebook.com/intern/diff/D18729561/)

[ghstack-poisoned]

albanD commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -764,42 +764,85 @@ TEST(OperatorRegistrationTest, whenRegisteringBackendFallbackKernelAndCatchallKe
+
+TEST(OperatorRegistrationTest, whenRegisteringAutogradKernelWithRegularKernel_thenCanCallAutogradKernel) {
+  auto registrar = c10::RegisterOperators().op("_test::dummy(Tensor dummy) -> ()", c10::RegisterOperators::options()
+    .impl_unboxedOnlyKernel<decltype(nonautograd_kernel), nonautograd_kernel>(TensorTypeId::CPUTensorId)
```
Why not `&nonautograd_kernel` here and everywhere `nonautograd_kernel` is used?

albanD reviewed (approved):
lgtm

Is there a specific reason why you do all test with Unboxed methods? Is the logic the same for Boxed ones and so doesn't need to be tested?

smessmer committed:
Update on "Custom op autograd tests"


Re-enable them and write a few additional ones

Differential Revision: [D18729561](https://our.internmc.facebook.com/intern/diff/D18729561/)

[ghstack-poisoned]
